src/lib.rs

```diff
@@ -104,104 +126,115 @@ impl<D> EventSource for WaylandSource<D> {
-            }
-        }
+        flush_queue(&mut self.queue)?;
 
-        // ensure we are not waiting for messages that are already in the buffer.
-        Self::loop_callback_pending(&mut self.queue, &mut callback)?;
-        self.read_guard = Some(Self::prepare_read(&mut self.queue)?);
+        self.read_guard = self.queue.prepare_read();
+        match self.read_guard {
+            Some(_) => Ok(None),
+            // If getting the guard failed, that means that there are
+            // events in the queue, and so we need to handle the events instantly
+            // rather than waiting on an event in polling. We tell calloop this
+            // by returning Some here. Note that the readiness value is
+            // never used, so we just need some marker
+            None => Ok(Some((Readiness::EMPTY, self.fake_token.unwrap()))),
+        }
+    }
 
-        Ok(())
+    fn before_handle_events(&mut self, events: calloop::EventIterator<'_>) {
+        // It's important that the guard isn't held whilst process_events calls occur
+        // This can use arbitrary user-provided code, which may want to use the wayland
+        // socket For example, creating a Vulkan surface needs access to the
+        // connection
+        let guard = self.read_guard.take();
+        if events.count() > 0 {
+            // Read events from the socket if any are available
+            if let Some(guard) = guard {
+                if let Err(WaylandError::Io(err)) = guard.read() {
+                    // If some other thread read events before us, concurrently, that's an expected
+                    // case, so this error isn't an issue. Other error kinds do need to be returned,
+                    // however
+                    if err.kind() != io::ErrorKind::WouldBlock {
+                        // before_handle_events doesn't allow returning errors
+                        // For now, cache it in self until process_events is called
+                        self.stored_error = Err(err);
+                    }
+                }
+            }
+        }
     }
+}
 
-    fn post_run<F>(&mut self, _: F) -> calloop::Result<()>
-    where
-        F: FnMut((), &mut Self::Metadata) -> Self::Ret,
-    {
-        // Drop implementation of ReadEventsGuard will do cleanup
-        self.read_guard.take();
-        Ok(())
+fn flush_queue<D>(queue: &mut EventQueue<D>) -> Result<(), calloop::Error> {
+    if let Err(WaylandError::Io(err)) = queue.flush() {
+        if err.kind() != io::ErrorKind::WouldBlock {
+            // in case of error, forward it and fast-exit
+            log_error!("Error trying to flush the wayland display: {}", err);
+            return Err(err.into());
+        }
+        // WouldBlock error means the compositor could not process all
+        // our messages quickly. Either it is slowed
+        // down or we are a spammer. Should not really
+        // happen, if it does we do nothing and will flush again later
```

Comment:
This comment still seems a bit weird, should be above the `WouldBlock` error.

Reply:
This was the existing style, but I've moved it up